lua/dap/ui/variable-tree.lua

```diff
@@ -0,0 +1,246 @@
+
+  local v = state.line_to_variable[pos[1] - 1]
+
+  if v and v.variablesReference > 0 then
+    if state.expanded[v.variablesReference] then
+      state.expanded[v.variablesReference] = nil
+      update_variable_buffer(buf)
+      api.nvim_win_set_cursor(win, pos)
+    else
+      state.expanded[v.variablesReference] = true
+      if v.variables then
+        update_variable_buffer(buf)
+        api.nvim_win_set_cursor(win, pos)
+      else
+        local session = state.session
+        if not session then
+          return
+        end
+        if not session.stopped_thread_id then
+          return
+        end
+        session:request(
+          "variables",
+          {variablesReference = v.variablesReference},
+          function(_, response)
+            if response then
+              v.variables =
+                utils.map_to_key(
+                function(var)
+                  return var.name
+                end,
+                response.variables
+              )
+              update_variable_buffer(buf)
+            end
+            api.nvim_win_set_cursor(win, pos)
+          end
+        )
+      end
+    end
+  end
+end
+
+local function create_variable_buffer(buf, win, session, root_variables)
+  if not buf then
+    buf = api.nvim_create_buf(true, true)
+
+    api.nvim_buf_set_option(buf, "buftype", "nofile")
+    api.nvim_buf_set_option(buf, "readonly", true)
+    api.nvim_buf_set_option(buf, "filetype", "dap-variable-ui")
+  end
+  api.nvim_buf_set_keymap(
+    buf,
+    "n",
+    "<CR>",
+    "<Cmd>lua require('dap.ui.variable-tree').toggle_variable_expanded()<CR>",
+    {}
+  )
+
+  variable_buffers[buf] = {variables = root_variables, expanded = {}, session = session, line_to_variable = {}}
+  for _, v in pairs(root_variables) do
+    -- Is variable expandable?
+    if v.variablesReference > 0 then
+      variable_buffers[buf].expanded[v.variablesReference] = true
+      if not v.variables then
+        session:request(
+          "variables",
+          {variablesReference = v.variablesReference},
+          function(_, response)
+            if response then
+              v.variables =
+                utils.map_to_key(
+                function(var)
+                  return var.name
+                end,
+                response.variables
+              )
+              update_variable_buffer(buf, win)
+            end
+          end
+        )
+      end
+    end
+  end
+  update_variable_buffer(buf, win)
+end
+
+-- Variable (https://microsoft.github.io/debug-adapter-protocol/specification#Types_Variable)
+--- name: string
+--- value: string (since we're also accepting scopes here: also optional)
+--- type?: string
+--- presentationHint?: VariablePresentationHint
+--- evaluateName?: string
+--- variablesReference: number
+--- namedVariables?: number
+--- indexedVariables?: number
+--- memoryReference?: string
+function M.write_variables(buf, variables, line, column, win)
```

Comment:
Do you  intend for users to call this directly?

Reply:
It's a remain from the time it was a in a more low-level module that would only write in existing buffers.